cpu/esp32/periph/spi.c

```diff
@@ -0,0 +1,432 @@
+{
+    DEBUG("%s bus=%u\n", __func__, bus);
+
+    assert(bus < SPI_NUMOF_MAX);
+    assert(bus < SPI_NUMOF);
+
+    /* initialize pins */
+    spi_init_pins(bus);
+
+    /* check whether pins could be initialized, otherwise return, CS is not
+       initialized in spi_init_pins */
+    if (gpio_get_pin_usage(spi_config[bus].sck) != _SPI &&
+        gpio_get_pin_usage(spi_config[bus].miso) != _SPI &&
+        gpio_get_pin_usage(spi_config[bus].mosi) != _SPI &&
+        gpio_get_pin_usage(spi_config[bus].cs) != _SPI) {
+        return;
+    }
+
+    /* enable (power on) the according SPI module */
+    esp_idf_periph_module_enable(_spi[bus].periph->module);
+
+    /* initialize SPI peripheral */
+    spi_ll_master_init(_spi[bus].periph->hw);
+
+    /* bring the bus into a defined state (one-line mode) */
+    spi_ll_master_set_line_mode(_spi[bus].periph->hw, (spi_line_mode_t){ 1, 1, 1 });
+    spi_ll_set_rx_lsbfirst(_spi[bus].periph->hw, false);
+    spi_ll_set_tx_lsbfirst(_spi[bus].periph->hw, false);
+
+    /* acquire and release to set default parameters */
+    spi_acquire(bus, GPIO_UNDEF, SPI_MODE_0, SPI_CLK_100KHZ);
+    spi_release(bus);
+
+    return;
+}
+
+void spi_init_pins(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+
+    /* avoid multiple pin initializations */
+    if (_spi[bus].pins_initialized) {
+        return;
+    }
+    _spi[bus].pins_initialized = true;
+
+    DEBUG("%s bus=%u\n", __func__, bus);
+
+    if (gpio_init(spi_config[bus].sck, GPIO_OUT) ||
+        gpio_init(spi_config[bus].mosi, GPIO_OUT) ||
+        gpio_init(spi_config[bus].miso, GPIO_IN)) {
+        LOG_TAG_ERROR("spi",
+                      "SPI_DEV(%d) pins could not be initialized\n", bus);
+        return;
+    }
+    if (spi_init_cs(bus, spi_config[bus].cs) != SPI_OK) {
+        LOG_TAG_ERROR("spi",
+                      "SPI_DEV(%d) CS signal could not be initialized\n",
+                      bus);
+        return;
+    }
+
+    /* store the usage type in GPIO table */
+    gpio_set_pin_usage(spi_config[bus].sck, _SPI);
+    gpio_set_pin_usage(spi_config[bus].mosi, _SPI);
+    gpio_set_pin_usage(spi_config[bus].miso, _SPI);
+
+    /* TODO  the IO_MUX should be used instead of GPIO matrix routing for
+             lower delays and higher clock rates whenever possible */
+
+    /* connect SCK and MOSI pins to the output signal through the GPIO matrix */
+    esp_rom_gpio_connect_out_signal(spi_config[bus].sck,
+                                    _spi[bus].periph->spiclk_out, false, false);
+    esp_rom_gpio_connect_out_signal(spi_config[bus].mosi,
+                                    _spi[bus].periph->spid_out, false, false);
+
+    /* connect MISO input signal to the MISO pin through the GPIO matrix */
+    esp_rom_gpio_connect_in_signal(spi_config[bus].miso,
+                                   _spi[bus].periph->spiq_in, false);
+}
+
+void spi_deinit_pins(spi_t bus)
+{
+    /* TODO */
+#if IS_USED(MODULE_SPI_RECONFIGURE)
```

Comment:
Might as well not advertise that feature, otherwise this would fail on CI

Reply:
It seems that using module `spi_reconfigure` requires feature `spi_reconfigure`. If this feature isn't defined, it should be possible to drop this function completely. In principle it is possible to reconfigure the pins on the ESP32x SoC, it is just not implemented yet and should not be part of this PR.